eras/alonzo/impl/src/Cardano/Ledger/Alonzo/Scripts.hs

```diff
@@ -610,21 +619,85 @@ decodeCostModel lang = do
 getEvaluationContext :: CostModel -> PV1.EvaluationContext
 getEvaluationContext (CostModel _ _ ec) = ec
 
-newtype CostModels = CostModels {unCostModels :: Map Language CostModel}
+-- | See 'CostModels' for an explanation of how 'CostModelError' is used.
+--
+-- It would be preferable to store the 'PV1.CostModelApplyError' type itself,
+-- but there is currently no 'Eq' instance for it, which we require.
+data CostModelError = CostModelError String
+  deriving stock (Eq, Show, Ord, Generic)
+
+instance NoThunks CostModelError
+
+-- | For a known version of Plutus, attempting to construct a cost model with
+-- too few parameters (depending on the version) will result in an error.
+-- 'CostModelApplyError' exists to collect these errors in the 'CostModels' type.
+-- The 'CostModels' type itself needs to be flexible enough to accept any map
+-- of 'Word8' to '[Integer]', so that cost models can be placed in the protocol parameters
+-- ahead of changes to the Plutus evaluation context. In this way, serializing a cost model,
+-- updating software, and deserializing can result in errors going away.
+--
+-- Additionally, 'CostModels' needs to be able to store cost models for future version
+-- of Plutus, which we cannot yet even validate. These are stored in 'invalidCostModels'.
+instance NFData CostModelError
+
+data CostModels = CostModels
+  { validCostModels :: Map Language CostModel
+  , costModelErrors :: Map Language CostModelError
+  , invalidCostModels :: Map Word8 [Integer]
+  }
   deriving stock (Eq, Show, Ord, Generic)
-  deriving newtype (NFData, NoThunks)
+
+emptyCostModels :: CostModels
+emptyCostModels = CostModels mempty mempty mempty
+
+-- | This function attempts to add a new cost model to a given 'CostModels'.
+-- If it is a valid cost model for a known version of Plutus, it is added to
+-- 'validCostModels'. If it is an invalid cost model for a known version of Plutus,
+-- the error is storted in 'costModelErrors' and the cost model is stored in
```

Comment:
`stored` instead of `storted`? 
However: can we make `storeted` a thing? to mean stored in a sorted way :grin: